### PR TITLE
#204 fixed, navigate statements

### PR DESCRIPTION
**Pull request by andreas-roehler**
New commands:

elixir-beginning-of-statement
elixir-beginning-of-statement-bol
elixir-end-of-statement
elixir-end-of-statement-bol
elixir-goto-statement-below

### CODE REVIEW

**Comment by tonini** [comment]
I like this :)

But I would like to go with ELisp conventions and naming functions with the prefix of the package itself.

like `elixir-intern-beginning-of-statement` or at least like the whole package itself `elixir-mode-beginning-of-statement`

What you think?

**Comment by mattdeboard** [comment]
Fixes #204

File: elixir-move.el [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    ;; reset
+    (setq elixir-bol-forms-last-indent nil)
+    (beginning-of-line)
+    (and (< (point) orig) (setq erg (point)))
+    (when (interactive-p) (message "%s" erg))
+    erg))
+
+(defalias 'elixir-forward-statement 'elixir-end-of-statement)
+(defalias 'elixir-previous-statement 'elixir-end-of-statement)
+(defalias 'elixir-statement-forward 'elixir-end-of-statement)
```
Comment (mattdeboard) [inline elixir-move.el]: What is the purpose of these aliases?

Comment (andreas-roehler) [inline elixir-move.el]: Am 27.05.2015 um 08:36 schrieb Matt DeBoard:
>
> In elixir-move.el 
> <https://github.com/elixir-lang/emacs-elixir/pull/205#discussion_r31106714>:
>
> > +    (if indent
> > +        (while (and
> > +                (setq erg (elixir-beginning-of-statement))
> > +                (< indent (current-indentation))
> > +                (not (bobp))))
> > +      (setq erg (elixir-beginning-of-statement)))
> > +    ;; reset
> > +    (setq elixir-bol-forms-last-indent nil)
> > +    (beginning-of-line)
> > +    (and (< (point) orig) (setq erg (point)))
> > +    (when (interactive-p) (message "%s" erg))
> > +    erg))
> > +
> > +(defalias 'elixir-forward-statement 'elixir-end-of-statement)
> > +(defalias 'elixir-previous-statement 'elixir-end-of-statement)
> > +(defalias 'elixir-statement-forward 'elixir-end-of-statement)
>
> What is the purpose of these aliases?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/elixir-lang/emacs-elixir/pull/205/files#r31106714>.
>

People coming from other modes might find the command via M-x apropos 
forward/previous/end RET likewise.

**Comment by andreas-roehler** [comment]
Am 27.05.2015 um 08:27 schrieb Samuel Tonini:
>
> I like this :)
>
> But I would like to go with ELisp conventions and naming functions 
> with the prefix of the package itself.
>
> like |elixir-intern-beginning-of-statement| or at least like the whole 
> package itself |elixir-mode-beginning-of-statement|
>
> What you think?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/elixir-lang/emacs-elixir/pull/205#issuecomment-105773145>.
>

Hmm, don't think there is such a convention. The only thing is a prefix 
- which must not be longer than needed. So would prefer "elixir" over 
"elixir-mode".

Package-name and prefix must not be the same.

Well, as "elixir-mode" is used already, can keep that.

What about changing it after merge?

**Comment by andreas-roehler** [comment]
Tests calling rake pass locally. Why does Travis CI fail? Functions claimed missing should be known via require.

**Comment by andreas-roehler** [comment]
Okay, see it. Some changes have not been checked in.

**Comment by andreas-roehler** [comment]
Am 27.05.2015 um 08:27 schrieb Samuel Tonini:
>
> I like this :)
>
> But I would like to go with ELisp conventions and naming functions 
> with the prefix of the package itself.
>
>

Uses "elixir-mode" now. Opened #209 for further discuss.

**Comment by tonini** [comment]
@andreas-roehler I really like what you did it's great :heart: 

But I'm a bit picky about some things. Could you please have a look at the following article by @nicferrier http://nic.ferrier.me.uk/blog/2012_07/tips-and-tricks-for-emacslisp

I think all the functions in `elixir-intern.el` should have the prefix `elixir-intern-` the functions which are just private should have a prefix `elixir-intern--` (for example `elixir-mode-preceding-line-backslashed-p` -> `elixir-intern--preceding-line-backslashed-p`)

After that I think we just could alias the functions `elixir-intern- beginning-of-statement` to `elixir-mode-beginning-of-statement`. Thats not a need at all, but it serves all the `elixir-mode` package interactive functions through the same place. For people who working with the minibuffer prompt it could be easier accessible to have all the `elixir-mode` package related functionality with the same prefix. 

What do you think?

/cc @mattdeboard

**Comment by andreas-roehler** [comment]
Will try to make as us like :)

Thanks for the link, there I read:

    Use -- to indicate private variables or functions, like:

(defun packagename--function

So private stuff would start elixir--whatever

Right? 

BTW "intern" indeed should come last, because doesn't convey much info.
Let's tell first, what a thing is about (and keep the obnoxious prefix short for that)

Cheers,

Andreas

**Comment by tonini** [comment]
> So private stuff would start elixir--whatever

I guess in our case it would be `elixir-intern--` because the package for navigating you made is called `elixir-intern`....

**Comment by andreas-roehler** [comment]
Mmm, see your point. So how to tell it isn't a package? A package designes some stuff belonging together.
In case these "intern" things, commoly that are subroutines, which must not have a relation to elixir-mode at all, not really. Did it into a separate file, because its simpler to re-use. 

For me emacs-elixir is the package, composed of files.

Mentioning the file-name in every indentifier would make it real hard to change/move things.

**Comment by tonini** [comment]
Maybe you're right.

I guess as a start I think we just should let it with just `elixir-` as you mentioned. :)

Could you just update the code that function which are private are also use that convention.

Then I think we could go with that :)

**Comment by andreas-roehler** [comment]
Thanks for your care. Will do some normalization: "--", move commands into ../move.el etc.; undo last commit and create a new branch.

**Comment by andreas-roehler** [comment]
Will be replaced by a new PR

**Pull request closed without merging**